pkg/batch/subscriber/processor/auditlog_persister.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/bucketeer-io/bucketeer/pkg/storage/v2/mysql"
+	domainevent "github.com/bucketeer-io/bucketeer/proto/event/domain"
+)
+
+type auditLogPersisterConfig struct {
+	FlushSize     int `json:"flushSize"`
+	FlushInterval int `json:"flushInterval"`
+}
+
+type auditLogPersister struct {
+	auditLogPersisterConfig auditLogPersisterConfig
+	mysqlAdminStorage       v2als.AdminAuditLogStorage
+	mysqlStorage            v2als.AuditLogStorage
+	logger                  *zap.Logger
+}
+
+func NewAuditLogPersister(
+	config interface{},
+	mysqlClient mysql.Client,
+	logger *zap.Logger,
+) (subscriber.Processor, error) {
+	auditLogPersisterJsonConfig, ok := config.(map[string]interface{})
+	if !ok {
+		logger.Error("AuditLogPersister: invalid config")
+		return nil, errAuditLogInvalidConfig
+	}
+	configBytes, err := json.Marshal(auditLogPersisterJsonConfig)
+	if err != nil {
+		logger.Error("AuditLogPersister: failed to marshal config", zap.Error(err))
+		return nil, err
+	}
+	var persisterConfig auditLogPersisterConfig
+	err = json.Unmarshal(configBytes, &persisterConfig)
+	if err != nil {
+		logger.Error("AuditLogPersister: failed to unmarshal config", zap.Error(err))
+		return nil, err
+	}
+	return &auditLogPersister{
+		auditLogPersisterConfig: persisterConfig,
+		mysqlAdminStorage:       v2als.NewAdminAuditLogStorage(mysqlClient),
+		mysqlStorage:            v2als.NewAuditLogStorage(mysqlClient),
+		logger:                  logger,
+	}, nil
+}
+
+func (a auditLogPersister) Process(
+	ctx context.Context,
+	msgChan <-chan *puller.Message,
+) error {
+	chunk := make(map[string]*puller.Message, a.auditLogPersisterConfig.FlushSize)
+	ticker := time.NewTicker(time.Duration(a.auditLogPersisterConfig.FlushInterval) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case msg, ok := <-msgChan:
+			if !ok {
+				return nil
+			}
+			subscriberReceivedCounter.WithLabelValues(subscriberAuditLog).Inc()
+			id := msg.Attributes["id"]
+			if id == "" {
+				msg.Ack()
+				subscriberHandledCounter.WithLabelValues(subscriberAuditLog, codes.MissingID.String()).Inc()
+				continue
+			}
+			if _, ok := chunk[id]; ok {
+				a.logger.Warn("Message with duplicate id", zap.String("id", id))
+				subscriberHandledCounter.WithLabelValues(subscriberAuditLog, codes.DuplicateID.String()).Inc()
+			}
+			chunk[id] = msg
+			if len(chunk) >= a.auditLogPersisterConfig.FlushSize {
+				a.flushChunk(chunk)
+				chunk = make(map[string]*puller.Message, a.auditLogPersisterConfig.FlushSize)
+			}
+		case <-ticker.C:
+			if len(chunk) > 0 {
+				a.flushChunk(chunk)
+				chunk = make(map[string]*puller.Message, a.auditLogPersisterConfig.FlushSize)
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (a auditLogPersister) flushChunk(chunk map[string]*puller.Message) {
+	auditlogs, adminAuditLogs, messages, adminMessages := a.extractAuditLogs(chunk)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
How about implementing the `flushInterval`?

Reply:
Do you mean that we re-implement `ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)` by using `flushInterval`?

Reply:
Yes.

Reply:
E.g.

```go
a.auditLogPersisterConfig.FlushTimeout
```